src/CommandLine.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_BASE_COMMANDLINE_H /* TODO: Move to src/base/ or remove BASE_. */
+#define SQUID_BASE_COMMANDLINE_H
+
+#include <getopt.h>
+#include <vector>
+
+/// Manages arguments passed to a program (i.e., main(argc, argv) parameters).
+class CommandLine
+{
+public:
+    /// expects main() input plus getopt_long(3) grammar rules for parsing argv
+    CommandLine(int argc, char *argv[], const char *shortRules, const struct option *longRules);
+    CommandLine(const CommandLine &them);
+    ~CommandLine();
+
+    // inefficient profile to simplify implementation
+    CommandLine &operator =(CommandLine);
+
+    /// \returns whether the option with optId identifier is present
+    /// When returning true, sets non-nil optValue to the found option's value.
+    /// For letter options (-x) and their --long synonyms, the letter is the ID.
+    /// For long-only --options, the ID is the configured options::val value.
+    bool hasOption(const int optId, const char **optValue = nullptr) const;
```

Comment:
hasOption() assignes optValue with `optarg`. This means that the next hasOption() call invalidates the result of the first one. Don't we need to document that?

Reply:
> the next hasOption() call invalidates the result of the first one

I do not think so. The option value a string stored in argv, which persists across method calls. `optarg` is a red herring in this context.

Reply:
that should probably be noted (inside the .cc for this call) for anyone else who is unaware of that hidden property.

Reply:
Done.